### PR TITLE
[FEATURE] Add PromQL metric finder to the explorer

### PR DESCRIPTION
**Pull request by Gladorme**
# Description

Add a metric finder to the metric explorer. Tested with a huge Thanos instance: it's possible to enable/disable metadata fetching, virtualized rendering, ...
Inspired by Promlens work ❤️ 

What could be implem in the futur: graph view and fuzzy search on lists
Currently dashboard variables inputs are not virtualized, I plan to use bring filters improvements to it

# Screenshots

image

image
image
image
image
image
image
image
image

image
image

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky.
  See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests)
  and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues
  include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

**Comment by Nexucis** [comment]
Hello from PromCon :)

I have shown a bit this preview about metrics finder and it returns that displaying graph for each metrics is not super interesting for a metrics finder. Specially for counter that should be used at least with a `rate` function before actually have something useful.
Also such view will hit the metrics platform.

I kind of agree with that statement also because it doesn't really work for Amadeus anyway. The list is working and is enough for the purpose of searching the metrics. Beside all the things you did when exploring the single metrics is really awesome and working really from the last the demo you shown.

So I would be in favor of removing the grid/graph view of all metrics

File: internal/cli/cmd/dac/build/testdata/go/go.mod [hunk 1]
```diff
@@ -13,9 +13,9 @@
 
 module dac
 
-go 1.22.4
+go 1.22.5
 
-toolchain go1.22.5
+toolchain go1.23.0
```
Comment (Nexucis) [inline internal/cli/cmd/dac/build/testdata/go/go.mod:18]: you can revert this change, it's not related to the PR ;)

Comment (Nexucis) [reply internal/cli/cmd/dac/build/testdata/go/go.mod:18]: that would solve the current conflict with the `main` branch

Comment (Gladorme) [reply internal/cli/cmd/dac/build/testdata/go/go.mod]: Yep, IntelliJ doing is stuff

File: ui/app/src/model/datasource-api.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ export class HTTPDatasourceAPI implements DatasourceApi {
    * correspond to nothing.
    * @param name
    * @param dashboard
-   * @param project
+   * @param project²
```
Comment (Nexucis) [inline ui/app/src/model/datasource-api.ts:32]: to be revert I suppose

Comment (Gladorme) [reply ui/app/src/model/datasource-api.ts:32]: yup

**Comment by Gladorme** [comment]
With the metadata, I can know if it's a counter (when provided) and use rate
image

**Comment by Nexucis** [comment]
And I know, doing cool stuff in UI can make it enjoyable.
But in this case, this view can hit the platform and put it in a bad shape. So I would say cool stuff are ok unless it impacts the infrastructure, the browser and basically the rest of the application (in term of performance, or simply it degrades the user experience)

**Comment by nicolastakashi** [comment]
Hey @Gladorme first of all amazing progress on that and this is really neat 🥳 
I tend to agree with @Nexucis in some aspects, maybe the list of panels would be a bit overloaded on very large setups, but I do like the graph while exploring a single metric.

**Comment by Gladorme** [comment]
@Nexucis @nicolastakashi The graph view can have all panels "disabled" by default and you coud click on a button to show it: it would have same perf as list. For now I removed it from the PR, but I think it can be a nice addition for small setups.

I still need to fix something with query params and the PR should be ready

**Comment by Nexucis** [comment]
cool :). I think once this PR is merged, we should create a new release. If I remember correctly @andreasgerstmayr is waiting for it since a long time now

**Comment by Gladorme** [comment]
@andreasgerstmayr FYI, I changed explorer params to be more usable for component not based on queries

**Comment by andreasgerstmayr** [comment]
> @andreasgerstmayr FYI, I changed explorer params to be more usable for component not based on queries

:+1:
Could you also update the link in the scatterchart: https://github.com/perses/perses/blob/c9389926c53f6d6151fb2c52726958f5811da8b9/ui/panels-plugin/src/plugins/scatterplot/ScatterChartPanel.tsx#L48-L54
Thanks!

I wonder if we should extract this code (generating the link) to some other common place, but I'm not sure where. We cannot import stuff from the explore component in the panels-plugin component.

File: ui/explore/src/components/PrometheusMetricsFinder/overview/MetricOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  isLoading?: boolean;
+}
+
+export function OverviewPanel({ metricName, datasource, filters, type, isLoading, ...props }: OverviewPanelProps) {
+  const { width, ref: panelRef } = useResizeObserver();
+  const suggestedStepMs = useSuggestedStepMs(width);
+
+  const { queries, definitions }: { queries: QueryDefinition[]; definitions: Array<Definition<UnknownSpec>> } =
+    useMemo(() => {
+      const expr =
```
Comment (AntoineThebaud) [inline ui/explore/src/components/PrometheusMetricsFinder/overview/MetricOverview.tsx:41]: Would be good to apply a rate() in the case of histograms & summaries also, when applicable:  even if most of the time they produce counter-like timeseries, it can also not be the case, ref https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations
Maybe a checkbox to enable/disable rate for these?

File: ui/explore/src/components/PrometheusMetricsFinder/display/MetricChip.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    return <Chip label={label} color="success" {...props} />;
+  }
+  if (label === 'counter') {
+    return <Chip label={label} color="primary" {...props} />;
+  }
+  if (label === 'histogram') {
+    return <Chip label={label} color="warning" {...props} />;
+  }
+  if (label === 'summary') {
+    return <Chip label={label} color="info" {...props} />;
```
Comment (AntoineThebaud) [inline ui/explore/src/components/PrometheusMetricsFinder/display/MetricChip.tsx:27]: Looks like color="info" for summary gives the exact same color as color="primary"?
image
 We could change to a different one for summaries for better differentiation

File: ui/explore/src/components/PrometheusMetricsFinder/utils.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        params['match[]'] = [`{${computeFilterExpr(filters)}}`];
+      }
+
+      return await client!.labelValues(params);
+    },
+  });
+}
+
+// Retrieve series from the Prometheus API for a given metric name and filters
+// Also computes the number of times a label value appears in a series
```
Comment (AntoineThebaud) [inline ui/explore/src/components/PrometheusMetricsFinder/utils.ts:108]:
```suggestion
// Also computes the number of times a label value appears for the given metric name and filters
```

**Review by AntoineThebaud — APPROVED** [review]
LGTM! GG for the hard work on this, I just got few remarks for stuff that could be tackled in a follow-up PR.

Besides below comments, I also spotted that:
- can be tedious to find a given metric when it has a long name since they are cropped on the display (well at least the tooltip is here to help):
2024-09-27_08h38_12
- histogram & summary metrics are mostly not supported/ showing unknown. I guess since these metric types are creating multiple series with different names it's trickier to retrieve the metadata here. Still it seems that Promlens manages to do it, so we may take a look at how it's achieved there.
E.g
image
vs 
image

File: ui/explore/src/components/PrometheusMetricsFinder/display/MetricChip.tsx (hunk 4 above)
Comment (Gladorme) [reply ui/explore/src/components/PrometheusMetricsFinder/display/MetricChip.tsx:27]: Yup
image

File: ui/explore/src/components/PrometheusMetricsFinder/overview/MetricOverview.tsx (hunk 3 above)
Comment (Gladorme) [reply ui/explore/src/components/PrometheusMetricsFinder/overview/MetricOverview.tsx:41]: Added a checkbox

**Review by Nexucis — APPROVED** [review]
awesome work @Gladorme. Thank you for doing it 🙏
